### PR TITLE
Add SD::BatchOptimizer::copy_from() function

### PR DESCRIPTION
**Pull request by jppelteret**

### CODE REVIEW

File: source/differentiation/sd/symengine_optimizer.cc [hunk 1]
```diff
@@ -63,10 +63,28 @@ namespace Differentiation
       , dependent_variables_output(0)
       , map_dep_expr_vec_entry(other.map_dep_expr_vec_entry)
       , ready_for_value_extraction(false)
+      , has_been_serialized(false)
```
Comment (bangerth) [inline source/differentiation/sd/symengine_optimizer.cc:66]: this is unrelated, right?

File: doc/news/changes/minor/20211127Jean-PaulPelteret [hunk 2]
```diff
@@ -0,0 +1,4 @@
+New: The SD::BatchOptimizer::copy_from() function can be used to duplicate the
+data from one batch optimizer instance to another.
+<br>
+(Jean-Paul Pelteret, 2021/11/27)
```
Comment (bangerth) [inline doc/news/changes/minor/20211127Jean-PaulPelteret:4]: add a newline to make github happy

File: source/differentiation/sd/symengine_optimizer.cc [hunk 3]
```diff
@@ -215,7 +233,7 @@ namespace Differentiation
 
     template <typename ReturnType>
     SD::types::symbol_vector
-    BatchOptimizer<ReturnType>::get_independent_symbols(void) const
+    BatchOptimizer<ReturnType>::get_independent_symbols() const
```
Comment (bangerth) [inline source/differentiation/sd/symengine_optimizer.cc:236]: Oops :-)

File: include/deal.II/differentiation/sd/symengine_optimizer.h [hunk 4]
```diff
@@ -1477,16 +1477,16 @@ namespace Differentiation
...
+       *
+       * @note The optimized data and results from previous substitutions
+       * executed by the @p other optimizer instance are not copied over.
+       * It is therefore necessary to re-optimize the data stored in
+       * this class, and it is possible to do so with a different optimization
+       * scheme.
        */
-      BatchOptimizer(const BatchOptimizer &other/*,
-                     const bool            copy_initialized = true*/);
+      BatchOptimizer(const BatchOptimizer &other);
```
Comment (bangerth) [inline include/deal.II/differentiation/sd/symengine_optimizer.h:1489]: I don't see where the corresponding change in the implementation is?

**Review by bangerth — CHANGES_REQUESTED** [review]

File: source/differentiation/sd/symengine_optimizer.cc (hunk 1 above)
Comment (jppelteret) [reply source/differentiation/sd/symengine_optimizer.cc]: Yeah, sorry about that. I didn't realise that this was a dirty commit. I've split out these cleanup changes into #13139.

**Review by bangerth — APPROVED** [review]
